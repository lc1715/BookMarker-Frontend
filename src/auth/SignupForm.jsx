import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import ShowAlert from '../common/ShowAlert';
import LoadSpinner from '../common/LoadSpinner';
//Material UI
import Container from '@mui/material/Container';
import Box from '@mui/material/Box';
import Paper from '@mui/material/Paper';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import OutlinedInput from '@mui/material/OutlinedInput';
import InputLabel from '@mui/material/InputLabel';
import InputAdornment from '@mui/material/InputAdornment';
import FormControl from '@mui/material/FormControl';
import Visibility from '@mui/icons-material/Visibility';
import VisibilityOff from '@mui/icons-material/VisibilityOff';


/** Signup form
 *
 * Route is '/signup'
 * 
 * - shows form
 * - updates the formData state when values in the input fields change
 * 
 * On submission:
 * - calls the signup function prop 
 * - redirects to homepage ('/')
 */

function SignupForm({ signup }) {
    /**Material UI's reveal/hide password related */
    const [showPassword, setShowPassword] = useState(false);
    const handleClickShowPassword = () => setShowPassword((show) => !show);
    const handleMouseDownPassword = (event) => { event.preventDefault() };
    const handleMouseUpPassword = (event) => { event.preventDefault() };
    /**End Material UI's reveal/hide password related */

    const [formData, setFormData] = useState({
        username: '',
        password: '',
        email: ''
    });
    const [formErrors, setFormErrors] = useState([]);
    const [loadingInfo, setLoadingInfo] = useState(false);
    const navigate = useNavigate();

    /** Handles signup form submission:
     * -Calls the signup function in App.jsx which will get user token
     * -If signup successful and token received, returns { success: true }, and navigates to homepage
     * -Otherwise returns { success: false, err }
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        setLoadingInfo(true);
        let result = await signup(formData);

        if (result.success) {
            navigate('/');
        } else {
            setFormErrors(result.err);
            setLoadingInfo(false);
        }
    }

    /**Update the formData state and show those changes in the input fields */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    }

    if (loadingInfo) {
        return <LoadSpinner />
    }

    return (
        <form onSubmit={handleSubmit}>
            <Container maxWidth="sm" sx={{
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                mt: { xs: 7, lg: 10 },
                mb: 4
            }}>
                <Paper elevation={4}
                    sx={{
                        width: { xs: '90%', lg: '85%' },
                        backgroundColor: '#FCFBF4'
                    }}>
                    <Box sx={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <Typography sx={{
                            fontSize: { xs: '35px', lg: '36px', xl: '40px' },
                            textAlign: 'center',
                            mt: 5
                        }}>
                            Signup Form
                        </Typography>
                    </Box>

                    <Box sx={{ width: '100%', mt: { xs: 3, xl: 2 }, display: 'flex', justifyContent: 'center' }}>
                        <TextField
                            name="username"
                            value={formData.username}
                            onChange={handleChange}
                            label="Username"
                            sx={{ width: { xs: '75%', lg: '73%' } }}
                        />
                    </Box>

                    <Box sx={{ width: '100%', mt: { xs: 2, lg: 1.5 }, display: 'flex', justifyContent: 'center' }}>
                        <FormControl sx={{ m: 1, width: { xs: '75%', lg: '73%' } }} variant="outlined">
                            <InputLabel htmlFor="password">Password</InputLabel>
                            <OutlinedInput
                                id="password"
                                name="password"
                                value={formData.password}
                                onChange={handleChange}
                                label="Password"
                                type={showPassword ? 'text' : 'password'}
                                endAdornment={
                                    <InputAdornment position="end">
                                        <IconButton
                                            aria-label={showPassword ? 'hide password' : 'display password'}
                                            onClick={handleClickShowPassword}
                                            onMouseDown={handleMouseDownPassword}
                                            onMouseUp={handleMouseUpPassword}
                                            edge="end"
                                        >
                                            {showPassword ? <VisibilityOff /> : <Visibility />}
                                        </IconButton>
                                    </InputAdornment>
                                }
                            />
                        </FormControl>
                    </Box>

                    <Box sx={{ width: '100%', mt: { xs: 2, lg: 1.5 }, display: 'flex', justifyContent: 'center' }}>
                        <TextField
                            name="email"
                            value={formData.email}
                            onChange={handleChange}
                            label="Email"
                            sx={{ width: { xs: '75%', lg: '73%' } }}
                        />
                    </Box>

                    {formErrors.length ?
                        <Box sx={{ mt: 1, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                            <ShowAlert type='error' messages={formErrors} />
                        </Box>
                        :
                        null}

                    <Box sx={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <Button
                            variant="contained"
                            size="medium"
                            sx={{ width: { xs: '75%', lg: '73%' }, height: 54, mt: { xs: 3, lg: 2.5 }, mb: { xs: 6.5, sm: 7.9, lg: 8 }, backgroundColor: '#cf8d86' }}
                            onClick={handleSubmit}
                            type='submit'
                        >
                            Submit
                        </Button>
                    </Box>
                </Paper>
            </Container >
        </form >
    )
};

export default SignupForm;
