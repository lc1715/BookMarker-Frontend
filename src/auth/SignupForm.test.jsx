import { it, expect } from 'vitest';
import { screen, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import SignupForm from './SignupForm';

it('renders the Signup Form component', () => {
    render(
        <MemoryRouter>
            <SignupForm />
        </MemoryRouter>
    )
})

it('matches with snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <SignupForm />
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('should render the element containing the text Signup Form', () => {
    render(
        <MemoryRouter>
            <SignupForm />
        </MemoryRouter>
    )

    expect(screen.getByText('Signup Form')).toBeInTheDocument()
})