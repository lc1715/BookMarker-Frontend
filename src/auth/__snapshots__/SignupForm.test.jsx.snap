// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-qd6y3d-MuiContainer-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation4 css-18icxhl-MuiPaper-root"
        style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-3p7yuy-MuiTypography-root"
          >
            Signup Form
          </p>
        </div>
        <div
          class="MuiBox-root css-1p547bn"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r3:"
              id=":r3:-label"
            >
              Username
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r3:"
                name="username"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Username
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-104zu77"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r4:"
              id=":r4:-label"
            >
              Password
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r4:"
                name="password"
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Password
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-104zu77"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r5:"
              id=":r5:-label"
            >
              Email
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r5:"
                name="email"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Email
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-lwq0sl-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
